invenio_rdm_records/tokens/resource_access.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Resource access tokens API."""
+
+from datetime import datetime, timedelta
+
+import jwt
+from flask import current_app
+from invenio_oauth2server.models import Token
+
+from .errors import (
+    ExpiredTokenError,
+    InvalidTokenError,
+    InvalidTokenIDError,
+    MissingTokenIDError,
+)
+from .scopes import tokens_generate_scope
+
+
+class ResourceAccessToken:
+    """Resource access tokens."""
+
+    def __init__(self, token):
+        """Initialize RAT token."""
+        self.token = token
+
+    def validate_rat(self, record, filename):
+        """Decodes a JWT token's payload and signer and performs validation."""
+        # Retrieve token ID from "kid"
+        try:
+            headers = jwt.get_unverified_header(self.token)
+            access_token_id = headers.get("kid")
+        except jwt.InvalidTokenError:
+            raise InvalidTokenError()
+
+        if not access_token_id:
+            raise MissingTokenIDError()
+        if not access_token_id.isdigit():
+            raise InvalidTokenIDError()
+
+        access_token = Token.query.get(int(access_token_id))
+        if (
+            not access_token
+            or tokens_generate_scope.id not in access_token.scopes
+            or (
+                access_token.expires is not None  # expired PAT check
+                and access_token.expires < datetime.utcnow()
+            )
+        ):
+            raise InvalidTokenError()
+
+        try:
+            payload = jwt.decode(
+                self.token,
+                key=access_token.access_token,
+                algorithms=current_app.config.get(
+                    "RDM_RESOURCE_ACCESS_TOKENS_WHITELISTED_JWT_ALGORITHMS",
+                    ["HS256", "HS384", "HS512"],
+                ),
+                options={"require_iat": True},
+            )
+
+            token_lifetime = current_app.config.get(
+                "RDM_RESOURCE_ACCESS_TOKENS_JWT_LIFETIME", timedelta(minutes=30)
+            )
+            # Verify that the token is not expired based on its issue time
+            issued_at = datetime.utcfromtimestamp(payload["iat"])
+            if (issued_at + token_lifetime) < datetime.utcnow():
+                raise ExpiredTokenError()
+
+        except jwt.InvalidTokenError:
+            raise InvalidTokenError()
+
+        rat_singer = access_token.user
+        payload = payload["sub"]
+
+        valid_claim = (
+            rat_singer.id == record.parent.access.owned_by[0].owner_id
+            and payload["record_id"] == record["id"]
+            and payload["file"] == filename
+            and payload["access"] == "read"
+        )
```

Comment:
I think this is not necessary here since we're only doing "formal" verification of the JWT signature, expiration, and structure. In that fashion, maybe you just verify that `access` is `read`/`write`, etc. and that at least a `record_id` key exists.

Thus the `record` and `filename` parameters are not needed in this function (see comment above about when/where to verify this).